### PR TITLE
feat: coalesce GraphQL/REST schema refresh on catalog schema changes

### LINKED ISSUE #1153: Coalesce GraphQL / REST schema refresh on catalog schema changes
# Coalesce GraphQL / REST schema refresh on catalog schema changes

## Background

The GraphQL and REST external APIs subscribe to the system CDC stream and rebuild
their per-catalog schemas (`GraphQLManager#refreshCatalog`,
`RestManager#refreshCatalog`) in reaction to engine mutations. The current wiring
wastes work in two ways:

1. **Per-mutation refresh.** Both observers
   (`SystemGraphQLRefreshingObserver:96-104`, `SystemRestRefreshingObserver:101-108`)
   react to every individual `ModifyCatalogSchemaMutation` by calling
   `refreshCatalog`. A single warm-up session that defines an entity and adds
   20+ attributes / sortable indexes / references generates ~30 separate engine
   mutations, each triggering a full GraphQL Data API + Schema API rebuild. On a
   catalog with 50 entities, each rebuild costs hundreds of ms. The observable
   effect (the schema clients see) is the same after every rebuild except the
   last, so 29 of those rebuilds are pure waste.

2. **Per-commit `CatalogInstalledIntoLiveView`.**
   `Evita#replaceCatalogReference` is called on every transaction commit
   (`TransactionManager#propagateCatalogSnapshot:1218`), which currently emits
   `CatalogInstalledIntoLiveView(ALIVE)` from `Evita#notifyCatalogStateSettled`.
   The observers handle this event by also calling `refreshCatalog`. Result:
   **every commit** — including pure data commits with no schema change — rebuilds
   both GraphQL schemas. This was an unintended side effect of the
   #1151 work; the event was meant to fire on state-settlement transitions, not
   on every catalog version increment.

The right shape is to fire schema-rebuild events **only when the catalog schema
version actually changed**, and to gate `CatalogInstalledIntoLiveView` on real
state transitions.

## Goal

Replace per-mutation observer reactions with a single coalesced signal —
`HostSystemEvent.CatalogSchemaUpdated` — emitted exactly once when a catalog's
schema version increases on this host, regardless of whether the bump came from
a single engine mutation or from N mutations within a session / transaction.

## Scope

**In scope:**

- New `HostSystemEvent.CatalogSchemaUpdated(catalogName, newSchemaVersion, currentEngineVersion)` variant.
- Emit at WARMING_UP session close and at ALIVE catalog version installation.
- Gate the existing `CatalogInstalledIntoLiveView` emit in
  `Evita#replaceCatalogReference` on actual state transition.
- Rewire GraphQL / REST observers to react to the new event instead of per-mutation
  events.
- Drop `SetCatalogMutabilityMutation` → `refreshCatalog` calls in both observers
  (the schema is mutability-independent today and stable schemas are a feature
  for clients — confirmed by Johnny 2026-05-07).
- gRPC, GraphQL, REST plumbing for the new event variant (proto, converters,
  descriptors, JSON / GraphQL union mappings).
- Tests covering coalescing, no-op data-only commit, and the existing test
  suite remaining green.

**Out of scope:**

- Read-only-aware schema building (intentionally rejected — schema stability is
  preferred over endpoint hiding).
- Per-transaction countdown coalescing on the catalog stream (TransactionMutation
  carries `mutationCount` but that path is not what drives the GraphQL/REST
  refresh churn).
- Lab UI / observability surface changes — the new event is a system CDC event
  and downstream subscribers can opt in; no UI work in this PR.

### New event variant

```java
record CatalogSchemaUpdated(
    @Nonnull String catalogName,
    int newSchemaVersion,
    long currentEngineVersion
) implements HostSystemEvent { ... }
```

Lives next to `CatalogInstalledIntoLiveView` and `CatalogRemovedFromLiveView`
in `evita_api/.../cdc/HostSystemEvent.java`. Constructor validates
`catalogName != null/empty` and `newSchemaVersion >= 0`.

### Emit site #1 — WARMING_UP (session close)

`EvitaSession` already records `this.startCatalogSchemaVersion` at construction
(line 359). At the end of `executeTerminationSteps` (line 1898) — i.e. after
the termination callback has fired but before `closeSession` returns — compare:

```java
final int finalVersion = theCatalog.getSchema().version();
if (finalVersion > this.startCatalogSchemaVersion) {
    this.evita.notifyCatalogSchemaUpdated(theCatalog.getName(), finalVersion);
}
```

A single emit per closing session that actually changed the schema. Sessions
that perform only data writes (or read-only sessions) emit nothing.

### Emit site #2 — ALIVE (transaction commit installs new catalog version)

`TransactionManager#propagateCatalogSnapshot:1218` calls
`newCatalogVersionConsumer.accept(newCatalogVersion)`, wired to
`Evita::replaceCatalogReference`. Inside that method, capture the prior
catalog's schema version before the swap, then compare:

```java
private void replaceCatalogReference(@Nonnull Catalog catalog) {
    final CatalogContract prior = getEngineState()
        .getCatalogInstance(catalog.getName()).orElse(null);
    final int priorSchemaVersion = prior == null ? -1 : prior.getSchema().version();
    final CatalogState priorState = prior == null ? null : prior.getCatalogState();

    getEngineState().replaceCatalogReference(catalog);

    final int newSchemaVersion = catalog.getSchema().version();
    if (newSchemaVersion > priorSchemaVersion) {
        notifyCatalogSchemaUpdated(catalog.getName(), newSchemaVersion);
    }
    if (priorState != catalog.getCatalogState()) {
        notifyCatalogStateSettled(catalog.getName(), catalog.getCatalogState());
    }
}
```

Pure data-only commits emit nothing. Schema-affecting commits emit exactly one
`CatalogSchemaUpdated`. State-changing replacements emit exactly one
`CatalogInstalledIntoLiveView`.

### Observer rewiring (GraphQL + REST)

Both observers currently have a `ModifyCatalogSchemaMutation` branch that calls
`refreshCatalog`. Replace it: stop reacting to per-mutation
`ModifyCatalogSchemaMutation` for refresh purposes; instead react to
`CatalogSchemaUpdated`:

```java
} else if (event instanceof HostSystemEvent.CatalogSchemaUpdated schemaUpdated) {
    this.graphQLManager.refreshCatalog(schemaUpdated.catalogName());
}
```

Drop the `SetCatalogMutabilityMutation` branch entirely (it produces no schema
change today and the project decision is to keep schemas stable across mutability
toggles).

The existing `CatalogInstalledIntoLiveView` branch stays — it handles
register-on-first-appearance — but with the gating change above it now fires
only on real state transitions, not per-commit.

### evita_api

- `src/main/java/io/evitadb/api/requestResponse/cdc/HostSystemEvent.java` —
  add `CatalogSchemaUpdated` record variant; update `permits` clause.

### evita_engine

- `src/main/java/io/evitadb/core/Evita.java` —
  add `notifyCatalogSchemaUpdated(String, int)` helper; modify
  `replaceCatalogReference` to capture prior schema version + state and emit
  conditionally.
- `src/main/java/io/evitadb/core/cdc/SystemChangeObserver.java` — extend
  `processHostEvent` switch to dispatch `CatalogSchemaUpdated`.
- `src/main/java/io/evitadb/core/session/EvitaSession.java` — emit at the end
  of `executeTerminationSteps` when schema version increased; needs access to
  the `Evita` instance (already held as `this.evita`).

### evita_external_api_grpc

- `shared/src/main/resources/META-INF/io/evitadb/externalApi/grpc/GrpcChangeCapture.proto` —
  add `GrpcCatalogSchemaUpdated` message + new `oneof` case in
  `GrpcHostSystemEvent`.
- `shared/src/main/java/.../requestResponse/cdc/ChangeCaptureConverter.java` —
  two helpers: `toGrpcCatalogSchemaUpdated`, `toCatalogSchemaUpdated`; extend
  the host-event dispatch in both directions.

### evita_external_api_core

- `src/main/java/.../system/model/cdc/CatalogSchemaUpdatedDescriptor.java` —
  new descriptor mirroring the existing
  `CatalogInstalledIntoLiveViewDescriptor` / `CatalogRemovedFromLiveViewDescriptor`.
- `src/main/java/.../system/model/cdc/HostSystemEventDescriptor.java` —
  register the new variant in the union / permitted list.

### evita_external_api_graphql

- `src/main/java/.../graphql/api/system/builder/SystemGraphQLSchemaBuilder.java` —
  add the new variant to the `ChangeSystemCaptureBodyUnion`.
- `src/main/java/.../graphql/api/system/resolver/subscribingDataFetcher/ChangeSystemCaptureUntypedBodyDataFetcher.java` —
  resolve `__typename` for the new variant.
- `src/main/java/.../graphql/api/catalog/SystemGraphQLRefreshingObserver.java` —
  add `CatalogSchemaUpdated` branch; remove `ModifyCatalogSchemaMutation` and
  `SetCatalogMutabilityMutation` refresh branches.

### evita_external_api_rest

- `src/main/java/.../rest/api/system/SystemRestBuilder.java` — register the new
  descriptor.
- `src/main/java/.../rest/api/system/resolver/serializer/ChangeSystemCaptureSerializer.java` —
  serialize the new variant with its discriminator.
- `src/main/java/.../rest/api/catalog/SystemRestRefreshingObserver.java` —
  same observer rewire as GraphQL.

### Engine-level

- `SystemChangeObserverTest`:
  - **schema mutation in WARMING_UP session emits one `CatalogSchemaUpdated`** —
    open session, define entity, add 5 attributes, close. Assert exactly one
    `CatalogSchemaUpdated(name, finalVersion)` arrives on a HOST subscriber.
  - **read-only WARMING_UP session emits no `CatalogSchemaUpdated`** — open
    read-only session, perform queries, close. Assert no host event of the new
    type.
  - **schema-affecting transaction emits one `CatalogSchemaUpdated` at commit**
    — on an ALIVE catalog, run a transaction that adds 3 attributes; assert
    exactly one event after commit.
  - **data-only transaction emits no `CatalogSchemaUpdated`** — on an ALIVE
    catalog, upsert 100 entities; assert no `CatalogSchemaUpdated` and no
    `CatalogInstalledIntoLiveView`.
  - **`CatalogInstalledIntoLiveView` no longer fires per-commit** — regression
    guard for the gating change.

### gRPC

- `ChangeCaptureConverterHostEventTest` — round-trip of the new variant.
- `EvitaClientReadWriteTest` — integration: subscribe with `hostArea()`, do a
  schema mutation, assert `CatalogSchemaUpdated` arrives.

### GraphQL / REST

- `SystemGraphQLSubscriptionsFunctionalTest` — subscribing client receives the
  new variant with correct `__typename`.
- `SystemRestStreamingFunctionalTest` — same for the JSON serializer.

### Refresh-counting integration

- New test (could live in
  `evita_test/.../externalApi/graphql/api/system/...` or as a thin engine-level
  test on a stub manager) that counts `refreshCatalog` invocations across a
  session that does N schema mutations. Pre-change: N calls. Post-change: 1
  call. The point of the whole work — make this the load-bearing test.

## Acceptance criteria

1. After a WARMING_UP session that performs M schema mutations, the relevant
   `Manager#refreshCatalog` is called exactly **once** (down from M).
2. After a single transaction that performs M schema mutations on an ALIVE
   catalog, the relevant `Manager#refreshCatalog` is called exactly **once**
   (down from M).
3. After a data-only transaction commit, the relevant `Manager#refreshCatalog`
   is called **zero** times (down from one per commit).
4. The default-shape system CDC subscription (no criteria) sees no behavior
   change. `CatalogSchemaUpdated` flows only to subscribers that opted into
   the `HOST` area, identical to the existing `CatalogInstalledIntoLiveView`
   contract.
5. All existing tests in `SystemChangeObserverTest`,
   `HostSystemEventOrderingTest`, `BootTimeAutoUpgradeHostEventTest`,
   `EvitaClientReadWriteTest`, `SystemGraphQLSubscriptionsFunctionalTest`,
   `SystemRestStreamingFunctionalTest` remain green.

## Risks and notes

- **Schema version monotonicity.** The implementation assumes
  `Catalog#getSchema().version()` is monotonically non-decreasing across
  reference replacements on this host. Verify before relying on the `>`
  comparison — `Catalog.java:1721`, `TransactionManager.java:720,1165` and the
  schema-version-delta tracking all treat it that way, but write a defensive
  test for replay / corruption recovery edge cases.
- **Engine-state lock ordering.** Both emit sites must run under the same
  `engineStateLock` chain that the existing host-event emits use, so events
  land strictly after their causing mutation in the system CDC stream. Reuse
  `Evita#notifyCatalogSchemaUpdated` exactly as `notifyCatalogStateSettled` is
  written today.
- **Live-tail-only contract preserved.** Like the other host events,
  `CatalogSchemaUpdated` is not persisted and not replayable from `sinceVersion`.
- **Gating side effect.** Removing the per-commit
  `CatalogInstalledIntoLiveView` emit means external observers no longer see
  a "catalog still ALIVE" pulse on each commit. This was unintentional churn,
  not a load-bearing signal — but if any subscriber relied on it as a heartbeat,
  they should switch to the proper `GrpcHeartBeat` mechanism (already wired).

## Estimated size

~15-20 file edits, balanced across `evita_api`, `evita_engine`, the three
external-API modules, and `evita_test`. Complexity is symmetric to the
`HostSystemEvent` work shipped under issue #1151 — same pattern, one extra
event variant.

## Reference

- Originating discussion: 2026-05-07 follow-up to issue #1151
- Related: `/home/jno/.claude/plans/imperative-plotting-flurry.md` (parent plan)
- Related: `/home/jno/.claude/plans/imperative-plotting-flurry-followups.md`
  (other deferred items from #1151)

---

Follow-up to #1151.

### PR DESCRIPTION
**Pull request by novoj**
## Summary

Schema mutations against a catalog historically caused GraphQL and REST managers to rebuild their entire API schema **once per `ModifyCatalogSchemaMutation`** — a refresh storm proportional to the number of mutations in a single transaction or warming-up session.

This PR introduces a new coalesced host event that fires **exactly once** per logical schema change:

- **WARMING_UP** — emitted on `EvitaSession#executeTerminationSteps` if the catalog schema version advanced during the session
- **ALIVE** — emitted from `Evita#replaceCatalogReference` (commit pipeline) when the snapshot swap actually advanced the schema version

GraphQL and REST refreshing observers no longer rebuild on per-mutation engine captures; they react solely to `HostSystemEvent.CatalogSchemaUpdated`.

### Key changes

- **API**: New sealed variant `HostSystemEvent.CatalogSchemaUpdated(catalogName, newSchemaVersion, timestamp)` with defensive validation in the compact constructor.
- **Engine**:
  - `Evita#replaceCatalogReference` simplified to a single map lookup; emits the host event when `ExpandedEngineState#replaceCatalogReference` reports the schema advanced.
  - `ExpandedEngineState#replaceCatalogReference` now returns `boolean` (schema-advanced flag) and captures the prior schema version atomically inside the swap.
  - `EvitaSession#executeTerminationSteps` inlines the gating decision into the `finally` block.
  - WAL-replay completion no longer emits a separate schema-update event — the catalog is not yet in the live view at that point, so live observers cannot have observed the pre-replay schema (this prevented a double rebuild).
- **External APIs**: `CatalogSchemaUpdatedDescriptor` exposed across GraphQL, REST, and gRPC (`GrpcCatalogSchemaUpdated` proto + `GrpcHostSystemEvent` wiring).
- **Observers**: `SystemGraphQLRefreshingObserver` and `SystemRestRefreshingObserver` rewired to drive their refresh path off the host event.
- **Tests**: Mock-heavy GraphQL/REST observer tests trimmed to coalescing-specific contracts; full dispatch coverage now lives in the GraphQL/REST functional subscription tests, which exercise a real server with real subscribers.

## Test plan

- [x] Unit tests pass: `SystemChangeObserverTest`, `BootTimeAutoUpgradeHostEventTest`, `HostSystemEventTest`, `HostSystemEventOrderingTest`
- [x] Coalescing contract tests pass: `SystemGraphQLRefreshingObserverTest`, `SystemRestRefreshingObserverTest`
- [x] Functional dispatch tests pass: `SystemGraphQLSubscriptionsFunctionalTest`, `SystemRestStreamingFunctionalTest`
- [x] gRPC roundtrip tests pass: `ChangeCaptureConverterHostEventTest`

Closes #1153

### CODE REVIEW

File: evita_engine/src/main/java/io/evitadb/core/engine/ExpandedEngineState.java [hunk 1]
```diff
@@ -218,27 +218,42 @@ public boolean isReadOnly(@Nonnull String catalogName) {
...
 		}
+		// UnusableCatalog placeholder cannot read its schema (would throw); treat as -1 so a
+		// real-Catalog replacement always counts as a schema advance. Defensive against a
+		// programming-error path — production swaps here are always Catalog→Catalog from the
+		// commit pipeline.
+		final int priorSchemaVersion = currentCatalog instanceof Catalog priorAlive
+			? priorAlive.getSchema().version()
+			: -1;
+		currentCatalogRef.replaceCatalogReference(catalog);
+		return catalog.getSchema().version() > priorSchemaVersion;
```
[the remark this answers is not in the document]
Comment (novoj) [reply evita_engine/src/main/java/io/evitadb/core/engine/ExpandedEngineState.java:258]: Fair point — the JavaDoc overstated the atomicity guarantee. The prior schema-version read happens **before** the atomic swap inside `CatalogWrapper#replaceCatalogReference`, so under concurrent calls for the same catalog the read can be stale relative to whoever wins the swap. In practice the commit pipeline (`TransactionManager#propagateCatalogSnapshot`) serializes calls per catalog, so the race is not reachable. Going with Option 1: softened the JavaDoc to match reality and document the serialization assumption rather than reshaping the accumulator. Fixed in 1f6cf83.

File: evita_engine/src/main/java/io/evitadb/core/Evita.java [hunk 2]
```diff
@@ -1553,28 +1560,39 @@ public void assertActiveAndWritable() {
...
 
 		// notify callback that it's now a live snapshot
 		catalog.notifyCatalogPresentInLiveView();
 
-		// emit the host event so external observers (GraphQL/REST/gRPC) can register endpoints
-		// for the freshly-installed catalog without restart — see issue #1151
-		notifyCatalogStateSettled(catalog.getName(), catalog.getCatalogState());
+		if (schemaAdvanced) {
+			notifyCatalogSchemaUpdated(catalog.getName(), catalog.getSchema().version());
+		}
```
[the remark this answers is not in the document]
Comment (novoj) [reply evita_engine/src/main/java/io/evitadb/core/Evita.java:1595]: Decline — the host-event channel is brand new in this release cycle (introduced in #1151 and extended here in #1153), so there are no existing subscribers whose contract we'd be breaking. The new behavior IS the intended semantic: `CatalogInstalledIntoLiveView` fires on state transitions, `CatalogSchemaUpdated` fires on schema bumps, data-only commits emit nothing on the host stream (subscribers wanting per-commit signals use `ChangeCatalogCapture`, the proper data channel). The previous per-commit `notifyCatalogStateSettled` from `replaceCatalogReference` was a leaky abstraction — the catalog was already installed, so labelling the swap as an "install" event was semantically wrong. No JavaDoc clarification needed for a contract no external party has yet seen.

File: evita_engine/src/main/java/io/evitadb/core/Evita.java [hunk 3]
```diff
@@ -1614,6 +1632,36 @@ public void notifyCatalogStateSettled(
...
+	 * at ALIVE catalog version installation. Coalesces what was previously a per-mutation
+	 * `refreshCatalog` storm in GraphQL / REST observers.
+	 *
+	 * The event is host-local and live-tail only — it does NOT advance the engine version
+	 * counter; the snapshot version it carries is for correlation only.
+	 *
+	 * @param catalogName       name of the catalog whose schema version increased; never `null`
+	 * @param newSchemaVersion  the new (current) catalog schema version on this host; non-negative
+	 */
+	public void notifyCatalogSchemaUpdated(
```
[the remark this answers is not in the document]
Comment (novoj) [reply evita_engine/src/main/java/io/evitadb/core/Evita.java:1647]: Decline — `notifyCatalogSchemaUpdated` has cross-package callers in the same module: `EvitaSession` (in `io.evitadb.core.session`) and the `SystemChangeObserverTest` shutdown-defensive test (in `io.evitadb.core.cdc`). Java's package-private visibility doesn't span sub-packages, so narrowing would force test plumbing or reflection. The sibling `notifyCatalogStateSettled` is also `public` for the same reason, so keeping symmetry. The JavaDoc already documents the internal-use intent.

File: evita_engine/src/main/java/io/evitadb/core/session/EvitaSession.java [hunk 4]
```diff
@@ -1922,6 +1922,18 @@ private void executeTerminationSteps(
...
+			// this point, so downstream observers must be notified regardless of callback
+			// success. ALIVE sessions are owned by the transaction-commit path
+			// (`Evita#replaceCatalogReference`) so we skip them here to avoid double-emit;
+			// read-only and pure data sessions trivially fall through
+			// (newSchemaVersion == startCatalogSchemaVersion).
+			final int newSchemaVersion = theCatalog.getSchema().version();
+			if (theCatalog.getCatalogState() != CatalogState.ALIVE
+					&& newSchemaVersion > this.startCatalogSchemaVersion) {
+				this.evita.notifyCatalogSchemaUpdated(theCatalog.getName(), newSchemaVersion);
+			}
```
[the remark this answers is not in the document]
Comment (novoj) [reply evita_engine/src/main/java/io/evitadb/core/session/EvitaSession.java:1936]: Good catch — fixed. Narrowed `startCatalogSchemaVersion` from `long` to `int` to match the actual return type of `CatalogSchema#version()`. The field originated in `b700dc76cd` (feat #260, ~1y8m ago) — looks like a defensive-typing oversight rather than a real reason; `version()` has always been `int`. All three usage sites already deal in `int`, zero behavioral change. Fixed in 1f6cf83.

File: evita_external_api/evita_external_api_grpc/shared/src/main/resources/META-INF/io/evitadb/externalApi/grpc/GrpcChangeCapture.proto [hunk 5]
```diff
@@ -109,6 +109,8 @@ message GrpcHostSystemEvent {
     GrpcCatalogInstalledIntoLiveView catalogInstalled = 1;
     // Fires when a catalog is fully removed from the live view on this host.
     GrpcCatalogRemovedFromLiveView catalogRemoved = 2;
```
[the remark this answers is not in the document]
Comment (novoj) [reply evita_external_api/evita_external_api_grpc/shared/src/main/resources/META-INF/io/evitadb/externalApi/grpc/GrpcChangeCapture.proto:111]: Thanks 🙂 — yes, the `EVENT_NOT_SET` forward-compat path on the converter side was deliberate, exactly so older clients survive when newer servers emit unknown variants over the same oneof.
